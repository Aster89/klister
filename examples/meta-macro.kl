#lang "n-ary-app.kl"

(import (shift "n-ary-app.kl" 1))
(import (shift "quasiquote.kl" 1))

(define x 'foo)

(define-macros
  ((meta-macro1 (lambda (stx)
                  (syntax-case stx
                    ((list (_ macro-name var-name value))
                     (pure `(define-macros
                              ((,macro-name (lambda (stx)
                                              (pure (list-syntax ('define ',var-name ',value) stx)))))))))))
   (meta-macro2 (lambda (stx)
                  (syntax-case stx
                    ((list (_ macro-name))
                     (pure `(define-macros
                              ((,macro-name (lambda (stx)
                                              (syntax-case stx
                                                ((list (_ var-name value))
                                                 (pure (list-syntax ('define var-name value) stx)))))))))))))))

(meta-macro1 my-macro1 x 'bar)
(my-macro1)
(example x)  -- 'foo

(meta-macro2 my-macro2)
(my-macro2 x 'baz)
(example x)  -- 'baz
