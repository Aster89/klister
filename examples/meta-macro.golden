#[/Users/gelisam/working/haskell/klister/examples/meta-macro.kl:7.12-7.15]<foo> : Syntax
#[/Users/gelisam/working/haskell/klister/examples/meta-macro.kl:31.15-31.18]
 <baz> : Syntax
